File: PDOO-Civitas/CivitasJ/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import GUI.*;
/**
 *
 * @author ramonjtorres
 */
public class Jugador implements Comparable<Jugador>{
    
    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected boolean encarcelado;
    protected static int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida = 1000;
    protected static float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static float SaldoInicial = 7500;
    ArrayList<TituloPropiedad> propiedades;
    Sorpresa salvoconducto;
    
    Jugador(String nombre){
        this.nombre = nombre;
        this.saldo = SaldoInicial;
        this.numCasillaActual = 0;
        this.propiedades = new ArrayList();
        this.puedeComprar = true;
        this.encarcelado = false;
        this.salvoconducto = null;
    }
    
    protected Jugador(Jugador otro){
        this.nombre = otro.nombre;
        this.numCasillaActual = otro.numCasillaActual;
        this.encarcelado = otro.encarcelado;
        this.propiedades = otro.propiedades;
        this.puedeComprar = otro.puedeComprar;
        this.salvoconducto = otro.salvoconducto;
        this.saldo = otro.saldo;
    }
    
    boolean cancelarHipoteca(int ip){
        
        boolean result = false;
        
        if(this.encarcelado){
            
            return result;
        }

        if(this.existeLaPropiedad(ip)){
        
            TituloPropiedad propiedad = this.propiedades.get(ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar = this.puedoGastar(cantidad);
            
            if(puedoGastar){
            
                result = propiedad.cancelarHipoteca(this);
                
                if(result){
                
                    Diario.getInstance().ocurreEvento("El jugador" + this.nombre + "cancela la hipoteca de la propiedad" + ip);
                }
            }
        }
        
        return result;
    }
    
    int cantidadCasasHoteles(){
        
        int cantidad = 0;
        
        for(TituloPropiedad p:propiedades){
        
            cantidad += p.cantidadCasasHoteles();
        }
        
        return cantidad;
    }
    
    public int compareTo(Jugador otro){
    
        return new Float(otro.saldo).compareTo(this.saldo);
    }
    
    boolean comprar(TituloPropiedad titulo){
        
        boolean result = false;
        
        if(this.encarcelado){
        
            return result;
        }
        
        if(this.puedeComprar){
    
            float precio = titulo.getPrecioCompra();
            
            if(this.puedoGastar(precio)){
            
                result = titulo.comprar(this);
                
                if(result){
                
                    this.propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " compra la propiedad " + titulo.getNombre());
                }
            }
        }
        
        return result;
    }
    
    boolean construirCasa(int ip){
        
        boolean puedoEdificarCasa = false;
        boolean result = false;
        float precio = 0;
        
        if(this.encarcelado){
         
            return result;
        }
        
        if(this.existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = propiedades.get(ip);
            puedoEdificarCasa = this.puedoEdificarCasa(propiedad);
            precio = propiedad.getPrecioEdificar();
            
            if((this.puedoGastar(precio))&&(propiedad.getNumCasas()<Jugador.CasasMax)){
                
                puedoEdificarCasa = true;
            }
            
            if(puedoEdificarCasa){
                
                result = propiedad.construirCasa(this);
                Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " construye casa en la propiedad "+ ip);
            }
        }
        
        return result;
        
    }
    
    boolean construirHotel(int ip){
        
        boolean result = false;
        
        if(this.encarcelado){
        
            return result;
        }
        
        if(this.existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = this.propiedades.get(ip);
            boolean puedoEdificarHotel = this.puedoEdificarHotel(propiedad);
            float precio = propiedad.getPrecioEdificar();
            
            if(puedoEdificarHotel){
            
                result = propiedad.construirHotel(this);
                int casasPorHotel = this.getCasasPorHotel();
                
                propiedad.derruirCasas(casasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador " + nombre + " construye hotel en la propiedad " + ip);
            }
            
        }
        
        return result;
    }
    
    protected boolean debeSerEncarcelado(){
    
        if(this.encarcelado){
     
            return false;
        }
        else if(!this.tieneSalvoconducto()){
            return true;
        }
        else{
            this.perderSalvoConducto();
            Diario.getInstance().ocurreEvento("El jugador se libra de la carcel");
            return false;
        }
        
    }
    
    boolean enBancarrota(){
        return this.saldo < 0;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(this.debeSerEncarcelado()){
            
            this.moverACasilla(numCasillaCarcel);
            this.encarcelado = true;
            Diario.getInstance().ocurreEvento("El jugador ha sido encarcelado");
        }
        return this.encarcelado;
    }
    
    private boolean existeLaPropiedad(int ip){
        
        return ip < this.propiedades.size();
    }
    
    int getCasasMax(){
        return CasasMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    int getHotelesMax(){
        return HotelesMax;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    public int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    public float getSaldo(){
        return saldo;
    }
    
    boolean hipotecar(int ip){
        
        boolean result = false;
        
        if(this.encarcelado){
        
            return result;
        }
        
        if(this.existeLaPropiedad(ip)){
        
            TituloPropiedad propiedad = this.propiedades.get(ip);
            result = propiedad.hipotecar(this);
            
            if(result){
                Diario.getInstance().ocurreEvento("El jugador" + this.nombre + "hipoteca la propiedad" + ip);
            }
        }
        
        return result;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    public boolean modificarSaldo(float cantidad){
        this.saldo = this.saldo + cantidad;
        Diario.getInstance().ocurreEvento("Se ha modificado el saldo ahora tienes " + this.saldo);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if(this.encarcelado)
            return false;
        else{
            this.numCasillaActual = numCasilla;
            this.puedeComprar = false;
            Diario.getInstance().ocurreEvento("El jugador ha sido movido de casilla a la numero " + this.numCasillaActual);
            return true;
        }
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        if(this.encarcelado)
            return false;
        this.salvoconducto = sorpresa;
        return true;
    }
    
    boolean paga(float cantidad){
        return this.modificarSaldo(-1*cantidad);
    }
    
    boolean pagaAlquiler(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.paga(cantidad);
    }
    
    boolean pagaImpuesto(float cantidad){
        if(this.encarcelado || cantidad == -1)
            return false;
        else
            return this.paga(cantidad);
    }
    
    boolean pasaPorSalida(){
        this.modificarSaldo(1000);
        Diario.getInstance().ocurreEvento("El jugador ha pasado por salida");
        return true;
    }
    
    private void perderSalvoConducto(){
        this.salvoconducto.usada();
        this.salvoconducto = null;
    }
    
    boolean puedeComprarCasilla(){
        if(this.encarcelado){
            
            this.puedeComprar = false;
        }
        else{
          
            this.puedeComprar = true;
        }
        
        return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        return (this.saldo>this.getPrecioLibertad());
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        
        if(this.propiedades.contains(propiedad) && this.saldo >= propiedad.getPrecioEdificar() && propiedad.getNumCasas() < this.getCasasMax()){
        
            return true;
        }
        else{
        
            return false;
        }
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        
        if(this.propiedades.contains(propiedad) && this.saldo >= propiedad.getPrecioEdificar() && propiedad.getNumCasas() == this.getCasasPorHotel() && propiedad.getNumHoteles() < this.getHotelesMax()){
        
            return true;
        }
        else{
        
            return false;
        }
    }
    
    boolean puedoGastar(float precio){
        if(this.encarcelado)
            return false;
        else
            return(this.saldo>precio);
    }
    
    boolean recibe(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.modificarSaldo(cantidad);
    }
    
    boolean salirCarcelPagando(){
        if(this.encarcelado){
         
            this.paga(this.getPrecioLibertad());
            this.encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel pagando");
            return true;   
        }
        else{
        
            return false;
        }
    }
    
    boolean salirCarcelTirando(){
        
        if(Dado.getInstance().salgoDeLaCarcel()){
            this.encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel tirando el dado");
            return true;
        }
        else return false;
    }
    
    boolean tieneAlgoQueGestionar(){
        
        if(this.propiedades.isEmpty()){
         
            return false;
        }
        else{
            
            return true;
        }
    }
    
    boolean tieneSalvoconducto(){
        return (this.salvoconducto != null);
    }
    
    boolean vender(int ip){
        
        if(this.encarcelado){
            
            return false;
        }
        else{
            
            if(this.existeLaPropiedad(ip)){
                
                if(propiedades.get(ip).vender(this)){
                 
                    Diario.getInstance().ocurreEvento("Se ha vendido la propiedad: " + propiedades.get(ip).toString());
                    propiedades.remove(ip);
                    return true;
                }
            }
            
            return false;
        }
    }
    
    @Override
    public String toString(){
     
        String s = "No";
        
        if(this.tieneSalvoconducto()){
        
            s = "Sí";
        }
        
        return "Nombre " + this.getClass() + ": " + this.getNombre() +
               "\nSaldo: " + this.getSaldo() +
               "\nCasilla Actual: " + this.getNumCasillaActual() +
               "\nSalvoconducto: " + s +
               "\nEncarcelado: " + this.isEncarcelado() +
               "\nPuede comprar: " + this.getPuedeComprar() +
               "\nPropiedades: " + "\n" +
                propiedades.toString(); 

    }
}

File: PDOO-Civitas/CivitasR/lib/jugador.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador
    include Comparable
    
    attr_accessor :encarcelado, :nombre, :num_casilla_actual, :propiedades, :puede_comprar, :saldo, :salvoconducto
    
    @@casas_max = 4
    @@hoteles_max = 4
    @@casas_por_hotel = 4
    @@paso_por_salida = 1000
    @@precio_libertad = 200
    @@saldo_inicial = 7500
    
    public
    def initialize(nombre, otro)
      
      if(otro == nil)
      
        @encarcelado = false
        @num_casilla_actual = 0
        @puede_comprar = true
        @saldo = @@saldo_inicial
        @nombre = nombre
        @propiedades = Array.new
        @salvoconducto = nil
        
      else
        
        @encarcelado = otro.encarcelado
        @num_casilla_actual = otro.num_casilla_actual
        @puede_comprar = otro.puede_comprar
        @saldo = otro.saldo
        @nombre = otro.nombre
        @propiedades = otro.propiedades
        @salvoconducto = otro.salvoconducto
        
      end
    end
    
    public
    def self.new_jugador(nombre)
      new(nombre, nil)
      
    end
    
    public
    def self.convertir_jugador(otro)
      new("",otro)
      
    end
    
    public
    def self.get_casas_max
      
      return @@casas_max
    end
    
    public
    def self.get_hoteles_max
      
      return @@hoteles_max
    end
    
    public
    def self.get_casas_por_hotel
      
      return @@casas_por_hotel
    end
    
    public
    def self.get_paso_por_salida
      
      return @@paso_por_salida
    end
    
    public
    def self.get_precio_libertad
      
      return @@precio_libertad
    end
    
    public
    def cancelar_hipoteca(ip)
      
      result = false
        
      if(@encarcelado)

          return result
      end

      if(existe_la_propiedad(ip))

          propiedad = @propiedades.at(ip)
          cantidad = propiedad.get_importe_cancelar_hipoteca()
          puedogastar = puedo_gastar(cantidad)

          if(puedogastar)

              result = propiedad.cancelar_hipoteca(self)

              if(result)

                  Diario.instance.ocurre_evento("El jugador " + @nombre + " cancela la hipoteca de la propiedad " + ip.to_s)
              end
          end
      end

      return result
      
    end
    
    public
    def cantidad_casas_hoteles()
      
      cantidad = 0
      
      @propiedades.each do |p|
        
            cantidad += p.cantidad_casas_hoteles();
      end
      
      return cantidad
      
    end
    
    public
    def <=>(otro)
      
      return (otro.saldo <=> @saldo)
    end
    
    public
    def comprar(titulo)
      
      result = false

      if(@encarcelado)

          return result
      end

      if(@puede_comprar)

          precio = titulo.precio_compra

          if(puedo_gastar(precio))

              result = titulo.comprar(self)

              if(result)

                  @propiedades.push(titulo)
                  Diario.instance.ocurre_evento("El jugador " + self.nombre + " compra la propiedad " + titulo.to_s)
              end
          end
      end

      return result
    end
    
    public
    def construir_casa(ip)
      result = false
      puedo_edificar_casa = false
      precio = 0
      if(@encarcelado)
        return result
      end
      
      if(existe_la_propiedad(ip))
        
        propiedad = @propiedades.at(ip)
        result = puedo_edificar_casa(propiedad)
        precio = propiedad.precio_edificar
        
        if((puedo_gastar(precio)) && (propiedad.num_casas < self.class.get_casas_max))
          puedo_edificar_casa = true
        end
        
        if(puedo_edificar_casa)
        
          result = propiedad.construir_casa(self)
          Diario.instance.ocurre_evento("El jugador "+ @nombre + " construye casa en la propiedad "+ ip.to_s)    
        end
      end
      
      return result;
    end
    
    public
    def construir_hotel(ip)
      
      result = false
        
      if(@encarcelado)
        
        return result
      end
        
        if(existe_la_propiedad(ip))
        
            propiedad = @propiedades.at(ip)
            puedoEdificarHotel = puedo_edificar_hotel(propiedad)
            precio = propiedad.precio_edificar
            
            if(puedoEdificarHotel)
            
                result = propiedad.construir_hotel(self)
                casasPorHotel = Jugador.get_casas_por_hotel()
                
                propiedad.derruir_casas(casasPorHotel, self)
                
                Diario.instance.ocurre_evento("El jugador " + @nombre + " construye hotel en la propiedad " + ip.to_s)
            end
            
        end
        
        return result
      
    end
    
    public
    def debe_ser_encarcelado()
      if(@encarcelado)
        return false
      elsif(!tiene_salvoconducto())
        return true
      else
        perder_salvoconducto()
        Diario.instance.ocurre_evento("El jugador se libra de la carcel")
        return false
      end
      
    end
    
    public
    def en_bancarrota()
      
      return @saldo < 0      
    end
    
    public
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado())
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        
        Diario.instance.ocurre_evento("El jugador ha sido encarcelado")
      end
      return @encarcelado
      
    end
    
    public
    def existe_la_propiedad(ip)
      
      return ip < @propiedades.length()
      
    end
    
    public
    def hipotecar(ip)
      
      result = false

      if(@encarcelado)

          return result
      end

      if(existe_la_propiedad(ip))

          propiedad = @propiedades.at(ip)
          result = propiedad.hipotecar(self)

          if(result)

              Diario.instance.ocurre_evento("El jugador " + @nombre + " hipoteca la propiedad " + ip.to_s)
          end
      end

      return result
    end
    
    public
    def is_encarcelado()
      
      return @encarcelado
    end
    
    public
    def modificar_saldo(cantidad)
      
      diario = Diario.instance
      
      @saldo = @saldo + cantidad
      diario.ocurre_evento("Se ha modificado el saldo ahora tienes "+@saldo.to_s)
      return true
      
    end
    
    public
    def mover_a_casilla(num_casilla)
      if(@encarcelado)
        return false
      else
        diario = Diario.instance
        
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        diario.ocurre_evento("El jugador ha sido movido de casilla a la numero "+ @num_casilla_actual.to_s)
        return true
      end
      
    end
    
    public
    def obtener_salvoconducto(sorpresa)
      if(@encarcelado)
        return false
      end
      @salvoconducto = sorpresa
      return true
    end
    
    public
    def paga(cantidad)
      modificar_saldo(-1*cantidad)
    end
    
    public
    def paga_alquiler(cantidad)
      if(@encarcelado)
        return false
      else
        return paga(cantidad)
      end
    end
    
    public
    def paga_impuesto(cantidad)
      if(@encarcelado || cantidad == -1)
        return false
      else
        return paga(cantidad)
      end
    end
    
    public
    def pasa_por_salida()
      modificar_saldo(1000)
      Diario.instance.ocurre_evento("El jugador ha pasado por salida")
      return true
      
    end
    
    public
    def perder_salvoconducto()
      
      @salvoconducto.usada()
      @salvoconducto = nil
      
    end
    
    public
    def puede_comprar_casilla()
      if(@encarcelado)
        @puede_comprar = false
      else
        @puede_comprar = true
      end
      return @puede_comprar
      
    end
    
    public
    def puede_salir_carcel_pagando()
      return @saldo>get_precio_libertad()
      
    end
    
    public
    def puedo_edificar_casa(propiedad)
      
      if(@propiedades.include?(propiedad) && @saldo >= propiedad.precio_edificar && propiedad.num_casas < self.class.get_casas_max)
       
        return true
        
      else
        
        return false
        
      end
      
    end
    
    public
    def puedo_edificar_hotel(propiedad)
      
      if(@propiedades.include?(propiedad) && @saldo >= propiedad.precio_edificar && propiedad.num_casas == self.class.get_casas_max && propiedad.num_hoteles < self.class.get_hoteles_max)
       
        return true
        
      else
        
        return false
        
      end
      
    end
    
    public
    def puedo_gastar(precio)
        if(@encarcelado)
          return false
        else
          return (@saldo>precio)
        end
      
    end
    
    public
    def recibe(cantidad)
      if(@encarcelado)
        return false
      else
        return modificar_saldo(cantidad)
      end
      
    end
    
    public
    def salir_carcel_pagando()
      if(@encarcelado)
        paga(@@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugado ha salido de la carcel pagando")
      end
      
    end
    
    public
    def salir_carcel_tirando()
      if(Dado.instance.salgo_de_la_carcel())
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador ha salido de la carcel tirando el dado")
        return true
      end
    end
    
    public
    def tiene_algo_que_gestionar()
      if(@propiedades.empty?)
        return false
      end
      return true
      
    end
    
    public
    def tiene_salvoconducto()
      return (@salvoconducto != nil)
    end
    
    public
    def vender(ip)
      if(@encarcelado)
        return false
      else
        if(existe_la_propiedad(ip))
          
          if(@propiedades.at(ip).vender(self))
            @propiedades.delete_at(ip)
            Diario.instance.ocurre_evento("Se ha vendido la propiedad: "+@propiedades.at(ip).to_s)
            return true
          end
        end
      end
    end
    
    public
    def to_s
      
      s = "No"
        
      if(tiene_salvoconducto())
        
        s = "Sí"
      end
        
      return "Nombre " + self.class.to_s + ": "+ @nombre.to_s +
             "\nSaldo: " + @saldo.to_s +
             "\nCasilla Actual: " + @num_casilla_actual.to_s +
             "\nSalvoconducto: " + s +
             "\nEncarcelado: " + @encarcelado.to_s +
             "\nPuede comprar: " + @puede_comprar.to_s +
             "\nPropiedades: " + "\n" +
             @propiedades.to_s
      
    end
 
  end
  
end
